Validate user account value

rvykydal opened the pull request:
The PR was originally open on anaconda repository and has some history and discussion there: https://github.com/rhinstaller/anaconda/pull/5305

rvykydal commented:
@garret please see https://github.com/rhinstaller/anaconda/pull/5305#pullrequestreview-1737335987

garrett commented on src/components/users/Accounts.jsx:
```diff
@@ -63,14 +93,39 @@ const CreateAccount = ({
+            valid = null;
+        } else if (userAccount.length > 32) {
+            valid = false;
+            setUserAccountInvalidHint(_("The user name is too long"));
```
Order of preference for the account name is:

1. Account name (doesn't use "user"; this is good, as "user" is bad to show in interfaces)
2. Username (the actual word, according to English dictionaries)
3. User name (an uncommon variant, according to English dictionaries)

As both GNOME and KDE talk about "user" accounts and have a "name" field, I guess keeping with "User name" is probably OK enough for consistency, even if it's not ideal.

Therefore, this is fine.

garrett replied:
On the other hand, it would be good to mention what the cutoff is.

```suggestion
            setUserAccountInvalidHint(_("User names must contain fewer than 33 characters"));
```

rvykydal replied:
Updated

garrett commented on src/components/users/Accounts.jsx:
```diff
@@ -63,14 +93,39 @@ const CreateAccount = ({
+            setUserAccountInvalidHint(_("The user name is too long"));
+        } else if (reservedNames.includes(userAccount)) {
+            valid = false;
+            setUserAccountInvalidHint(_("Sorry, that user name is not available. Please try another."));
```
Since this is blocked exclusively because it is reserved word, it might be nice to say that specifically here.

```suggestion
            setUserAccountInvalidHint(_("This is a reserved word; please try another user name"));
```

rvykydal replied:
Updated

garrett commented on src/components/users/Accounts.jsx:
```diff
@@ -63,14 +93,39 @@ const CreateAccount = ({
+            setUserAccountInvalidHint(_("Sorry, that user name is not available. Please try another."));
+        } else if (isUserAccountWithInvalidCharacters(userAccount)) {
+            valid = false;
+            setUserAccountInvalidHint(cockpit.format(_("The user name should usually only consist of lower case letters from a-z, digits and the following characters: $0"), "-_"));
```
A few improvements:

- Rewording
- "lowercase" is one word
- Oxford comma for an inline list

```suggestion
            setUserAccountInvalidHint(cockpit.format(_("User names may only contain lowercase letters from a-z, digits, and the following characters: $0"), "-_"));
```

rvykydal replied:
Updated

garrett reviewed (changes requested):
I've made a bunch of suggestions.

garrett commented:
⚠️ Also, in the future, when you're tagging me on GitHub, please use my name without typos. When you make typos, it makes other people upset... and I also won't see the ping.

(My name has two Rs and two Ts.)

rvykydal commented:
> ⚠️ Also, in the future, when you're tagging me on GitHub, please use my name without typos. When you make typos, it makes other people upset... and I also won't see the ping.
> 
> (My name has two Rs and two Ts.)

Sorry for that, relying on auto-completion makes me ignorant, now I'll remember :)

rvykydal commented:
![Screenshot from 2023-11-21 14-40-50](https://github.com/rhinstaller/anaconda-webui/assets/1220237/9883286d-0f46-4b7b-85a9-74224a5803bf)
![Screenshot from 2023-11-21 14-41-12](https://github.com/rhinstaller/anaconda-webui/assets/1220237/cb74e8fe-6a7a-4e5f-9cb7-e081a867c21f)
![Screenshot from 2023-11-21 14-42-47](https://github.com/rhinstaller/anaconda-webui/assets/1220237/308ea3d4-a015-4a11-ac83-dd7243367d66)

garrett commented on src/components/users/Accounts.jsx:
```diff
@@ -63,14 +93,39 @@ const CreateAccount = ({
+            valid = null;
+        } else if (userAccount.length > 32) {
+            valid = false;
+            setUserAccountInvalidHint(_("User names must contain fewer than 33 characters"));
```
```suggestion
            setUserAccountInvalidHint(_("User name must be shorter than 33 characters"));
```

garrett replied:
```suggestion
            setUserAccountInvalidHint(_("User name must be shorter than 33 characters"));
```

garrett commented on src/components/users/Accounts.jsx:
```diff
@@ -63,14 +93,39 @@ const CreateAccount = ({
+            setUserAccountInvalidHint(_("User names must contain fewer than 33 characters"));
+        } else if (reservedNames.includes(userAccount)) {
+            valid = false;
+            setUserAccountInvalidHint(_("This is a reserved word; please try another user name"));
```
```suggestion
            setUserAccountInvalidHint(_("User name must not be a reserved word"));
```

garrett commented on src/components/users/Accounts.jsx:
```diff
@@ -63,14 +93,39 @@ const CreateAccount = ({
+            setUserAccountInvalidHint(_("This is a reserved word; please try another user name"));
+        } else if (isUserAccountWithInvalidCharacters(userAccount)) {
+            valid = false;
+            setUserAccountInvalidHint(cockpit.format(_("User names may only contain lowercase letters from a-z, digits, and the following characters: $0"), "-._"));
```
```suggestion
            setUserAccountInvalidHint(cockpit.format(_("User name may only contain: letters from a-z, digits 0-9, dash -, period ., and underscore _")));
```

garrett reviewed (changes requested):
Thanks for the screenshots!

I'm looking at this all again, in context now, with a clearer head (less of a post-COVID headache today), and I have another round of suggestions. Apologies!

This time around, I'll show high fidelity mockups and explain it:

![Create account, error long](https://github.com/rhinstaller/anaconda-webui/assets/10246/611d7361-2265-4b2a-a752-be8bbf047a68)

- Since we're talking about "user names", we should be consistent with the label.
- This is using PatternFly recommendations for errors in the forms, with the red helper text .
- I shortened the text to "User name must be shorter than 33 characters"
- Like the original mockups, the user name field is shorter, to show the limit visually. (It should roughly fit 33 wide characters, like 33 "w"s side by side, plus a little padding for the error icon.)

![Create account, error reserved](https://github.com/rhinstaller/anaconda-webui/assets/10246/47cdcc40-ca7a-4b95-8907-2b3338f3b08f)

- We don't need to say "please change" as that's implied
- I changed the text to be a little clearer
- I've added a (?) icon for a popover to explain it
- The popover shows the reserved words
- This icon and popover is a "nice to have" and should be a follow-up... do not block this PR for it :wink: 

![Create account, error character type](https://github.com/rhinstaller/anaconda-webui/assets/10246/dea7716b-71d5-415e-9b1d-93aefade0ea2)

- I changed the text to be shorter for the description and made it more consistent with the details (with the word first, then symbols after)

Again: Apologies for another review with different changes after you already addressed what I said yesterday. Thanks in advance!

I've also attached the changes as suggestions here, so I hope that makes it easier to apply. I did not see where the "User account" label is in this PR (so I could change it to "User name", to make it more consistent with the rest of the writing as well as with GNOME, KDE, and probably other desktops.)

rvykydal commented:
> 
>     * Since we're talking about "user names", we should be consistent with the label.
> 
>     * This is using PatternFly recommendations for errors in the forms, with the red helper text .
> 
>     * I shortened the text to "User name must be shorter than 33 characters"
> 
>     * Like the original mockups, the user name field is shorter, to show the limit visually. (It should roughly fit 33 wide characters, like 33 "w"s side by side, plus a little padding for the error icon.)
>

Modified. 33 "w"s seems maybe too wide to me (too close to the width of the other entries) but I am ok with it:

![Screenshot from 2023-11-22 17-15-33](https://github.com/rhinstaller/anaconda-webui/assets/1220237/05ad5914-a7a6-4a8f-99bd-5de07d3d0cbc)

![Screenshot from 2023-11-22 17-16-05](https://github.com/rhinstaller/anaconda-webui/assets/1220237/4493cd73-aed7-49a5-adc5-99b0c88b8ee9)




 

> 
>     * We don't need to say "please change" as that's implied
> 
>     * I changed the text to be a little clearer

Updated

> 
>     * I've added a (?) icon for a popover to explain it
> 
>     * The popover shows the reserved words
> 
>     * This icon and popover is a "nice to have" and should be a follow-up... do not block this PR for it 😉
> 

I'd leave it as eventual follow-up. 
(Personally I don't find the list very useful, IMO the change you hit the reserved name is quite low. Also there is a plan to move the validation to the backend so the backend would need to provide the list)

![Screenshot from 2023-11-22 17-16-17](https://github.com/rhinstaller/anaconda-webui/assets/1220237/be367522-95ad-4e8d-9eba-ce484e6210a7)




> 
>     * I changed the text to be shorter for the description and made it more consistent with the details (with the word first, then symbols after)
> 

Updated.
Just  a note - some rejected names falling under this message are not really consistent with the message (same as in gnome users where the original message comes from) but I think it is acceptable given they are pretty rare (maybe except for numbers-only). (https://github.com/rvykydal/anaconda-webui/blob/da5daf8533c5b030f8a78d36a0c1c8ec099aead7/src/components/users/Accounts.jsx#L81, https://github.com/rvykydal/anaconda-webui/blob/da5daf8533c5b030f8a78d36a0c1c8ec099aead7/test/check-users#L125)

![Screenshot from 2023-11-22 17-15-48](https://github.com/rhinstaller/anaconda-webui/assets/1220237/2974de75-41f2-424e-838c-d4324f13edf4)

garrett commented:
Thanks for the changes!

> Modified. 33 "w"s seems maybe too wide to me (too close to the width of the other entries) but I am ok with it:

Yeah, that does seem wide. I guess on a larger resolution screen, there might be more of a difference?

> I'd leave it as eventual follow-up.
> (Personally I don't find the list very useful, IMO the change you hit the reserved name is quite low. Also there is a plan to move the validation to the backend so the backend would need to provide the list)

Yeah, agreed on both points. It's pretty unlikely someone will hit multiples of the reserved words. Having it be a follow-up is totally fine. Forgetting to follow-up is probably OK too, as it's pretty unlikely.

> rejected names falling under this message are not really consistent with the message (same as in gnome users where the original message comes from) but I think it is acceptable given they are pretty rare (maybe except for numbers-only)

Should we have an error for numbers only then?

"User names must contain at least one letter from a-z."

Although I guess you could have a username of `$`, right? :thinking:

KKoukiou commented on src/components/users/Accounts.jsx:
```diff
@@ -55,6 +58,33 @@ export const accountsToDbusUsers = (accounts) => {
     }];
 };
 
+const reservedNames = [
```
This code now is duplicated pyanaconda/core/users.py

I will create an issue to expose it via an API and remove the duplication.

KKoukiou replied:
https://issues.redhat.com/browse/INSTALLER-3839

rvykydal commented:
> Thanks for the changes!
> 
> > Modified. 33 "w"s seems maybe too wide to me (too close to the width of the other entries) but I am ok with it:
> 
> Yeah, that does seem wide. I guess on a larger resolution screen, there might be more of a difference?
>

I don't think so, actually the width of other entries is already limited by limiting the form width: https://github.com/rvykydal/anaconda-webui/blob/da5daf8533c5b030f8a78d36a0c1c8ec099aead7/src/components/users/Accounts.scss#L3.
 

> > rejected names falling under this message are not really consistent with the message (same as in gnome users where the original message comes from) but I think it is acceptable given they are pretty rare (maybe except for numbers-only)
> 
> Should we have an error for numbers only then?
> 
> "User names must contain at least one letter from a-z."
> 
> Although I guess you could have a username of `$`, right? 🤔

If anything, I'd say something like "numbers-only user name is not allowed"  because for example those are valid: `3.4`, `4$`. (But I am fine with the current PR state)
